modules/dhcp_common/free_ips.rb

```diff
@@ -0,0 +1,133 @@
+require 'concurrent'
+require 'logger'
+
+module Proxy::DHCP
+  class FreeIps
+    include ::Proxy::Log
+
+    DEFAULT_ADDRESS_KEEP_TIME = 30*60 # 30 mins
+    DEFAULT_CLEANUP_INTERVAL = 60 # 1 min
+
+    attr_reader :allocated_ips, :allocation_timestamps, :m
+
+    def initialize
+      @allocated_ips = []
+      @allocation_timestamps = []
+      @m = Monitor.new
+    end
+
+    # must be called from under the monitor
+    def mark_ip_as_allocated(ip_address)
+      @allocated_ips.push(ip_address)
+      @allocation_timestamps.push(time_now + DEFAULT_ADDRESS_KEEP_TIME)
+    end
+
+    def clean_up_allocated_ips
+      @m.synchronize do
+        loop do
+          logger.debug("Starting allocated ip addresses cleanup pass...")
+          break if @allocation_timestamps.first.nil? || @allocation_timestamps.first > time_now
+          @allocation_timestamps.shift
+          ip = @allocated_ips.shift
+          logger.debug("#{ip} marked as free.")
```

Comment:
I'd appreciate compacting the two debug statements into single line (IPs joined by comma).

Reply:
They are separate on purpose -- the first one indicates the cleanup started, the second that something has been cleaned up (which may not always be the case).